[SYCL] Basic code generation for SYCL kernel caller offload entry point functions.

tahonermann opened the pull request:
A function declared with the `sycl_kernel_entry_point` attribute, sometimes called a SYCL kernel entry point function, specifies a pattern from which the parameters and body of an offload entry point function, sometimes called a SYCL kernel caller function, are derived.

SYCL kernel caller functions are emitted during SYCL device compilation. Their parameters and body are derived from the `SYCLKernelCallStmt` statement and `OutlinedFunctionDecl` declaration associated with their corresponding SYCL kernel entry point function. A distinct SYCL kernel caller function is generated for each SYCL kernel entry point function defined as a non-inline function or ODR-used in the translation unit.

The name of each SYCL kernel caller function is parameterized by the SYCL kernel name type specified by the `sycl_kernel_entry_point` attribute attached to the corresponding SYCL kernel entry point function. For the moment, the Itanium ABI mangled name for typeinfo data (`_ZTS<type>`) is used to name these functions; a future change will switch to a more appropriate naming scheme.

The calling convention used for a SYCL kernel caller function is target dependent. Support for AMDGCN, NVPTX, and SPIR targets is currently provided. These functions are required to observe the language restrictions for SYCL devices as specified by the SYCL 2020 specification; this includes a forward progress guarantee and prohibits recursion.

Only SYCL kernel caller functions, functions declared as `SYCL_EXTERNAL`, and functions directly or indirectly referenced from those functions should be emitted during device compilation. Pruning of other declarations has not yet been implemented.

elizabethandrews committed:
[SYCL] Basic code generation for SYCL kernel caller offload entry point functions.

A function declared with the `sycl_kernel_entry_point` attribute, sometimes
called a SYCL kernel entry point function, specifies a pattern from which the
parameters and body of an offload entry point function, sometimes called a
SYCL kernel caller function, are derived.

SYCL kernel caller functions are emitted during SYCL device compilation.
Their parameters and body are derived from the `SYCLKernelCallStmt` statement
and `OutlinedFunctionDecl` declaration associated with their corresponding
SYCL kernel entry point function. A distinct SYCL kernel caller function is
generated for each SYCL kernel entry point function defined as a non-inline
function or ODR-used in the translation unit.

The name of each SYCL kernel caller function is parameterized by the SYCL
kernel name type specified by the `sycl_kernel_entry_point` attribute attached
to the corresponding SYCL kernel entry point function. For the moment, the
Itanium ABI mangled name for typeinfo data (`_ZTS<type>`) is used to name
these functions; a future change will switch to a more appropriate naming
scheme.

The calling convention used for a SYCL kernel caller function is target
dependent. Support for AMDGCN, NVPTX, and SPIR targets is currently provided.
These functions are required to observe the language restrictions for SYCL
devices as specified by the SYCL 2020 specification; this includes a forward
progress guarantee and prohibits recursion.

Only SYCL kernel caller functions, functions declared as `SYCL_EXTERNAL`, and
functions directly or indirectly referenced from those functions should be
emitted during device compilation. Pruning of other declarations has not yet
been implemented.

Co-authored-by: Tom Honermann <tom.honermann@intel.com>

Fznamznon commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,177 @@
+// and emited during device compilation. They are not emitted during device
+// compilation.
+
+template <typename name, typename Func>
```
a NIT
```suggestion
template <typename Name, typename Func>
```

Fznamznon replied:
Also, does it make sense to introduce a fake sycl.hpp header like done in the downstream? This will keep tests aligned and the test cases will be closer to the real case scenarios.

tahonermann replied:
It probably will eventually, but I prefer to keep tests that exercise very low level functionality like this as bare as possible.

tahonermann replied:
Name style inconsistency fixed, thank you! I switched to using the names that are used in the SYCL specification. I also switched to using C++11 -style attribute syntax because I find it easier to visually parse.

Fznamznon commented on clang/lib/CodeGen/CodeGenTypes.h:
```diff
@@ -229,6 +229,12 @@ class CodeGenTypes {
                                                    const CallArgList &args);
 
+  /// A SYCL device kernel function is a free standing function with
+  /// spir_kernel calling convention
```
```suggestion
  /// spir_kernel calling convention.
```

bader replied:
```suggestion
  /// target specific calling convention (e.g. spir_kernel for the SPIR target).
```
Technically we use OpenCL language calling convention, but I'm not sure if we should specify that in the comments. Typically, each offload target has its own calling convention. For example, OpenCL compiler emits `amdgpu_kernel` calling convention for AMD GPU targets and `ptx_kernel` for NVIDIA GPU targets.

NOTE: Setting calling convention might be not enough for decorating kernel functions properly. E.g. NVPTX target uses named metadata to identify kernel functions. 
https://docs.nvidia.com/cuda/nvvm-ir-spec/index.html#representation-of-properties
I suppose this means that SYCL compiler for NVPTX target must emit named metadata referring to the kernel in addition to using `ptx_kernel` calling convention.
Some SPIR tools consuming LLVM IR doesn't require additional metadata, so setting calling convention is enough. I don't know how NVPTX backend will handle LLVM w/o the metadata.
@npmiller, FYI.

tahonermann replied:
Thank you for the comment correction; I'll fix that. This change does produce the calling convention tags you noted (see the IR for the `kernel-caller-entry-point.cpp` test below). The changes to emit the kernel metadata will be included in the next PR; I deferred it from this one to ease code review.

tahonermann replied:
I fixed the comment, thank you!

Thank you for mentioning the kernel metadata and linking those previous PRs. As I mentioned, generation of kernel metadata is planned for the next PR I'll submit. Should it continue to emit the `kernel` annotation for NVPTX targets to align with the NVVM IR Specification? The code that we have in development currently does not do so.

tahonermann replied:
Answering my own question here. https://github.com/llvm/llvm-project/pull/120806 removed all the calls that caused the `kernel` NVVM metadata to be emitted as illustrated by both code and test changes. I checked for other calls in both LLVM/Clang and in the oneAPI sources and didn't find any. Resolving this comment now.

Fznamznon commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,177 @@
+// RUN: %clang_cc1 -fsycl-is-host -emit-llvm -triple x86_64-pc-windows-msvc %s -o - | FileCheck --check-prefixes=CHECK-HOST,CHECK-HOST-WINDOWS %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple amdgcn %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-AMDGCN %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple nvptx %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-NVPTX %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple spir64 %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-SPIR %s
```
Does it make sense to test with SPIR-V backend too?

tahonermann replied:
Sure, I'll add that. There don't seem to be any SYCL tests that exercise the spirv target upstream and very few in the oneAPI sources. I guess we should check 32-bit and 64-bit targets as well.

tahonermann replied:
I added additional `nvptx64`, `spir`, `spirv32`, and `spirv64` targets. I also expanded the triples for all targets to match the full triple used by other tests.

Fznamznon commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -732,6 +732,16 @@ CodeGenTypes::arrangeBuiltinFunctionDeclaration(CanQualType resultType,
+const CGFunctionInfo &
+CodeGenTypes::arrangeSYCLKernelCallerDeclaration(QualType resultType,
+                                                 const FunctionArgList &args) {
+  auto argTypes = getArgTypesForDeclaration(Context, args);
```
Will this place eventually create more changes to kernel signature in order to support sycl accessor type?

tahonermann replied:
No, the effects of parameter decomposition should already have been realized and be reflected in the `OutlinedFunctionDecl` declaration created for the `sycl_kernel_entry_point` attributed function. The commit message for commit https://github.com/llvm/llvm-project/commit/8fb42300a02c887740825cd1b60fc4fcd8d2f933 might be helpful to better explain the design intent.

Fznamznon commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3303,6 +3303,27 @@ void CodeGenModule::EmitDeferred() {
+          FD->isDefined()) {
+        // Functions with an invalid sycl_kernel_entry_point attribute are
+        // ignored during device compilation.
+        if (!FD->getAttr<SYCLKernelEntryPointAttr>()->isInvalidAttr()) {
```
Can we simply drop the attribute if it is invalid when checking its validity?

tahonermann replied:
If the attribute is syntactically invalid, it won't get added to the function. Attributes that are semantically invalid are preserved and marked like this so that errors can be properly emitted during overload resolution, particularly for instantiated function template specializations. These attributes should have no effect on which overload candidate is selected during overload resolution, but might be instantiated during that process. Errors that occur during instantiation should not be issued unless the candidate is actually selected by overload resolution. It is therefore useful to retain these attributes so that error messages can be produced at the right time. This isn't working exactly as it should yet though; see [these tests](https://github.com/llvm/llvm-project/blob/2ca27e7c3e252d5c5fce57c0f85d0290babf82f3/clang/test/SemaSYCL/sycl-kernel-entry-point-attr-sfinae.cpp#L30-L65).

tahonermann reviewed (commented):
Thank you for the review comments @Fznamznon and @bader! I added responses to some of the comments and will get updates applied tomorrow.

tahonermann committed:
Updates to address code review feedback from Mariya Podchishchaeva and Alexey Bader.

- Corrected a misleading comment.
- Corrected inconsistent naming style in a test.
- Expanded a test to exercise 32-bit and 64-bit SPIR, SPIR-V, and NVPTX.

tahonermann commented:
@Fznamznon, @frasercrmck, any further review feedback? @erichkeane, would you like to weigh in on this change before it lands?

frasercrmck commented:
Nothing from me!

erichkeane commented:
You managed to submit this while I was on vacation, so I'll take a look at this in the next few days.

erichkeane commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,186 @@
+// attributed functions during host compilation. ODR-use of these functions may
+// require them to be emitted, but they have no effect if called.
+//
+// CHECK-HOST-LINUX:      define dso_local void @_Z26single_purpose_kernel_task21single_purpose_kernel() #[[LINUX_ATTR0:[0-9]+]] {
```
What is the point with the named capture, if it isn't being used anywhere?

erichkeane commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,186 @@
+// CHECK-HOST-LINUX:      define dso_local void @_Z26single_purpose_kernel_task21single_purpose_kernel() #[[LINUX_ATTR0:[0-9]+]] {
+// CHECK-HOST-LINUX-NEXT: entry:
+// CHECK-HOST-LINUX-NEXT:   %kernelFunc = alloca %struct.single_purpose_kernel, align 1
+// CHECK-HOST-LINUX-NEXT:   ret void
```
I don't really get what is happening here... and I wonder if we should be inserting an llvm.trap here instead?  Since this should never be called.

erichkeane commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,186 @@
+// CHECK-HOST-LINUX-NEXT:   ret void
+// CHECK-HOST-LINUX-NEXT: }
+//
+// CHECK-HOST-WINDOWS:      define dso_local void @"?single_purpose_kernel_task@@YAXUsingle_purpose_kernel@@@Z"(i8 %kernelFunc.coerce) #[[WINDOWS_ATTR0:[0-9]+]] {
```
Same question on attributes here.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -12816,6 +12816,15 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
+    if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLKernelEntryPointAttr>())
+      return true;
+
+    // FIXME: Functions declared with SYCL_EXTERNAL are required during
```
would love a test that shows that they aren't currently being emitted, with a FIXME that says it should.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14794,9 +14803,36 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+  auto DeviceDiscriminatorOverrider =
+      [](ASTContext &Ctx, const NamedDecl *ND) -> std::optional<unsigned> {
+    if (const auto *RD = dyn_cast<CXXRecordDecl>(ND))
+      if (RD->isLambda())
```
What happens with `isGenericLambda`? It seems that this could cause issues if we are willing to emit/choose a number for a generic lambda, right?  (note that `isLambda` is true for generic lambdas).

tahonermann replied:
Why do you believe that device mangling numbers are assigned differently for non-generic lambdas vs generic lambdas? I don't see any reason that they would be treated differently. Can you point to relevant code? The device mangling number applies to the (single) closure type produced by a (generic) lambda.
> Also, it doesn't seem that we are properly transforming the device mangling number by my investigation of the code.

Can you elaborate? I don't know what you're referring to. This code is nearly identical to `SYCLUniqueStableNameExpr::ComputeName()`.

tahonermann replied:
Sorry, but I need a more concrete description of the concern to know what I'm looking for. I audited uses of `isGenericLambda()` that I thought might be relevant and I didn't find anything that suggested relevance for device mangling context. Can we postpone this until the upcoming PR that revises the name mangling approach? That PR will come with tests and I can ensure that generic lambdas are exercised.

erichkeane replied:
My concern here is: What happens if a `genericLambda` gets here, which I believe will have an unset 'mangling-id'.   I'd VERY much like to not let this get pushed to the next one, in part since the ways forward aren't a particularly sizable amount of work.

I see a few ways forward:
1- Create an example that shows the mangler properly handling the generic-lambda.  Note you might have to do some work to get here, as getting generic-lambdas mangled often requires some surprising things to make happen.
2- Prove in some-way that generic-lambda cannot get here (which, would actually take quite a bit of convincing, dependent types find their way to sneak into the mangler all the time unfortunately, particularly with SFINAE and `diagnose_if`/etc).
2- 'assert(!isGenericLambda() && "dont' know how to handle generic lambdas yet");`

tahonermann replied:
I modified the test included with this PR to use a generic lambda as the kernel name type. The mangled name is reflected in the mangled name used for the kernel entry point. I hope that assuages your concerns. If not, I'll need a theoretical explanation for why the capture type of a generic lambda would be mangled any differently than for a non-generic lambda in order to make progress. I can't think of any reason for mangling differences.

Given your mention of `TransfromLambdaExpr` earlier, is it possible you are thinking of a case of a dependent lambda expression as opposed to a generic lambda?

tahonermann replied:
@erichkeane,
> Is this what you're referring to?

Yes.

> If so, I think that does.

Excellent.

> The concern here is that you're using DeviceLambdaManglingNumber, which, at one point when I checked, wasn't actually set for non-instantiated generic lambdas. So the attempt to mangle them would always result in the 'same' mangling number, which doesn't really work. In addition to the above, can you test the following, and make sure it results in proper looking names?

I modified the test (locally) as follows. The use of `decltype` and variables is necessary since lambda expressions aren't allowed in contexts that would require mangling.
```
  auto lambda = [=](auto) { (void) capture; };
  auto lx2 = [](int){};
  auto lx3 = [](auto){};  // Note, same signature as 'lambda'.
  auto lx4 = [](auto, auto){};
  kernel_single_task<decltype(lambda)>(lambda);
  kernel_single_task<decltype(lx2)>(lambda);
  kernel_single_task<decltype(lx3)>(lambda);
  kernel_single_task<decltype(lx4)>(lambda);
```

That produced the following names for the kernel entry points (with their associated demangling):
```
_ZTSZ4mainEUlT_E_
typeinfo name for main::'lambda'(auto)
_ZTSZ4mainEUliE_
typeinfo name for main::'lambda'(int)
_ZTSZ4mainEUlT_E0_
typeinfo name for main::'lambda0'(auto)
_ZTSZ4mainEUlT_T0_E_
typeinfo name for main::'lambda'(auto, auto)
```

I didn't commit this change because this test is intended to validate generated code, not the names. An extensive test for names will come in the PR that implements our new mangling strategy and it will include additions like those above.

> Depending on how types are mangled in some situations we actually mangle the arguments of the lambda, which results in auto being used as the mangling, though my memory on when that happens is vague ATM.

The argument signature is always included in the mangling of the closure type as that reduces the need for discriminators. This goes for both generic and non-generic lambdas as shown above (only the lambda corresponding to `lx3` gets a discriminator bump).

I looked through `clang/lib/AST/ItaniumMangle.cpp` for anything that is sensitive to generic lambdas. There are two such locations. `CXXNameMangler::mangleLambda()` calls `CXXNameMangler::mangleLambdaSig()` to incorporate the lambda signature; this includes any requires clauses on a generic lambda. The other is in `CXXNameMangler::TemplateArgManglingInfo::isOverloadable()` and is used to inform how template arguments are mangled.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14794,9 +14803,36 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+
+  // Construct a mangled name for the SYCL kernel caller offload entry point.
+  // FIXME: The Itanium typeinfo mangling (_ZTS<type>) is currently used to
+  // FIXME: name the SYCL kernel caller offload entry point function. This
```
the FIXME on every line for a FIXME isn't consistent with the rest of the project, and frankly, is absurdly confusing to read.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14794,9 +14803,36 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+  // FIXME: name the SYCL kernel caller offload entry point function. This
+  // FIXME: mangling does not suffice to clearly identify symbols that
+  // FIXME: correspond to SYCL kernel caller functions, nor is this mangling
+  // FIXME: natural for targets that use a non-Itanium ABI.
```
What alternatives do we have here?  Can we discuss briefly in-source what the 'appropriate' fix to this is?   Should I expect this in a followup patch, or do we expect this in 'the future' as an ABI break?

erichkeane commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -732,6 +732,16 @@ CodeGenTypes::arrangeBuiltinFunctionDeclaration(CanQualType resultType,
+const CGFunctionInfo &
+CodeGenTypes::arrangeSYCLKernelCallerDeclaration(QualType resultType,
+                                                 const FunctionArgList &args) {
+  auto argTypes = getArgTypesForDeclaration(Context, args);
```
Not a place we should use `auto` I believe.

tahonermann replied:
I don't think these uses violate the LLVM coding standard. The relevant guidelines are linked below.
- https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
- https://llvm.org/docs/CodingStandards.html#beware-unnecessary-copies-with-auto

It looks to me like `getArgTypesForDeclaration()` follows NRVO, so copies should be avoided.

erichkeane replied:
Aaron's interpretation of that for `auto` is "if the name of the type is on the RHS of the equals" (such as in casts/etc).  So I don't think the type of `argTypes` is there, unless I'm missing it?

tahonermann replied:
The type is not on the RHS. I'm being consistent with the existing code. If you would like the code changed to introduce a type alias for the returned type and the current uses of `auto` replaced with that alias, please submit a PR or file a GH issue for someone to make that change. I don't consider it in scope for this PR.

erichkeane replied:
The existence of existing code violating our coding standard (as interpreted by our maintainer) is NOT sufficient justification to ignore it in new code.  In fact, our policy is the opposite.  It is NOT acceptable to 'make things worse' in this way.

AaronBallman replied:
IMO, the coding standard says to spell out the type here. However, I think you should spell out of the type not because a rule told you to but because a reviewer asked for it (that alone should suffice). But as a matter of practicality: should that be `const auto &` instead? Does the type have a move constructor? These sort of questions come up for reviewers and we shouldn't have to chase down types we can't see.

FWIW, another option is to get rid of `argTypes` entirely and nest the function call instead. The local variable isn't really giving much value to begin with.

> I don't consider it in scope for this PR.

This is new code, it's in scope for this PR to adjust the way you express it.

tahonermann replied:
@AaronBallman,
> FWIW, another option is to get rid of argTypes entirely and nest the function call instead. The local variable isn't really giving much value to begin with.

In general, such transformations have order of evaluation implications and the lack of a variable makes debugging more difficult. But yes, it is an option.

> >  I don't consider it in scope for this PR.
> 
> This is new code, it's in scope for this PR to adjust the way you express it.

That comment was in reference to the previous sentence regarding introducing a type alias and changing the existing uses of `auto`.

Given two maintainers have expressed displeasure at the use of `auto` here, I'll introduce a type alias for the return types of `getArgTypesForCall()`, `getArgTypesForDeclaration()`, and `getExtParameterInfosForCall()` and replace the existing uses of `auto` with the alias. Were I just to spell out the type name here, I think the inconsistency with surrounding code would be worse than using `auto` (my personal opinion is that consistency with surrounding code is more important than strict coding style conformance; no one else is obligated to agree of course).

tahonermann replied:
@erichkeane, @AaronBallman, I submitted a separate PR to address the existing uses of `auto`. Once that lands, I'll rebase this PR. https://github.com/llvm/llvm-project/pull/135861.

tahonermann replied:
The rebase lost the context for this conversation thread, so it is no longer found on the "Files changed" tab; it can still be found in the "Conversation" tab though.

This use of `auto` has now been replaced with `CanQualTypeList` and consistent style with surrounding code is retained (due to the rebase on top of the changes for #135861). Resolving this conversation.

erichkeane commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -732,6 +732,16 @@ CodeGenTypes::arrangeBuiltinFunctionDeclaration(CanQualType resultType,
+
+  return arrangeLLVMFunctionInfo(
+      GetReturnType(resultType), FnInfoOpts::None, argTypes,
+      FunctionType::ExtInfo(CC_OpenCLKernel), {}, RequiredArgs::All);
```
I'm not familiar with what the `paramInfos` arg is supposed to be doing here.  Can you clarify why you don't need anything there?

erichkeane commented on clang/lib/CodeGen/CodeGenSYCL.cpp:
```diff
@@ -0,0 +1,73 @@
+         " during device compilation");
+
+  const auto *KernelEntryPointAttr =
+      KernelEntryPointFn->getAttr<SYCLKernelEntryPointAttr>();
```
Do we do anything to make sure that you can't have multiples of this attribute on the same decl?  Either via inheritance (`InheritableAttr`) or by multiple definition?

erichkeane commented on clang/lib/CodeGen/CodeGenSYCL.cpp:
```diff
@@ -0,0 +1,73 @@
+  // Find the SYCLKernelCallStmt.
+  SYCLKernelCallStmt *KernelCallStmt =
+      dyn_cast<SYCLKernelCallStmt>(KernelEntryPointFn->getBody());
+  assert(KernelCallStmt && "SYCLKernelCallStmt not found");
```
Just do `cast` above, it has an assert in it already without the additional overhead of the dyn_cast/check here.

Fznamznon commented on clang/test/CodeGenSYCL/kernel-caller-entry-point.cpp:
```diff
@@ -0,0 +1,186 @@
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple spir-unknown-unknown %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-SPIR %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple spir64-unknown-unknown %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-SPIR %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple spirv32-unknown-unknown %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-SPIR %s
+// RUN: %clang_cc1 -fsycl-is-device -emit-llvm -aux-triple x86_64-pc-windows-msvc -triple spirv64-unknown-unknown %s -o - | FileCheck --check-prefixes=CHECK-DEVICE,CHECK-SPIR %s
```
I was kind of expecting that we will be checking SPIR-V instructions with that, but I don't really know how SPIR-V backend is being tested, so you may ignore this.

bader replied:
`-emit-llvm` forces the compiler to emit LLVM IR instead of SPIR-V.
Compiler emits SPIR-V binary if `emit-llvm` option is not set.

tahonermann replied:
Is it worth adding a test to validate SPIR-V (and PTX) output? I'm inclined not to since such the transformation from LLVM IR to other formats is a backend concern that should already be exercised elsewhere for each of the LLVM IR features present in this test.

Fznamznon reviewed (approved):
I don't think I have anything to add to the existing feedback.

tahonermann committed:
Updates to address code review feedback from Erich Keane.

- Changed the style of added FIXME comments.
- Changed a `dyn_cast` followed by `assert` to `cast`.
- Removed some unused `FileCheck` captures in a test.

tahonermann reviewed (commented):
Thank you for the review @erichkeane! I believe I have addressed or responded to each of your comments.

tahonermann committed:
Addressed clang-format complaints.

tahonermann commented:
@erichkeane, in case you missed the previous notice, this is ready for your continued review.

erichkeane commented:
> @erichkeane, in case you missed the previous notice, this is ready for your continued review.

Thanks, I did miss that.  I'll try to get to this later this week.

tahonermann commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3303,6 +3303,27 @@ void CodeGenModule::EmitDeferred() {
+          EmitSYCLKernelCaller(FD, getContext());
+          // Recurse to emit any symbols directly or indirectly referenced
+          // by the SYCL kernel caller function.
+          EmitDeferred();
```
I misunderstood your question. We can't continue because we don't need or want to emit `D` on the device side and there is no `GlobalDecl` object corresponding to the SYCL kernel caller function that we can substitute for it. The call to `EmitDeferred()` here substitutes for not reaching the call at line 3363.

erichkeane replied:
Ah, got it.  Thanks for the clarification!

erichkeane reviewed (changes requested):
See comment re:auto: our lead maintainer's interpretation of our `auto` rules requires that we not use `auto` in this situation.   @AaronBallman for attention.

tahonermann committed:
Updates to address code review feedback from Erich Keane.

- Test modified to use a generic lambda.

tahonermann committed:
Updates to address code review feedback from Erich Keane.

- Replaced use of `auto` with a type alias added via
  https://github.com/llvm/llvm-project/pull/135861.

tahonermann commented:
> I can't find our conversation again and firefox crashed before I could submit, but PLEASE make sure the mangling tests cover the examples we put together.

I absolutely will.

> ALSO, we don't always mangle the types of lambdas, we sometimes do $_0 and $_1 (source of another bug) for ones without external linkage.

Yup. That gets us into the fun topic of what it means for a symbol with internal linkage to be accessible from both a host and device TU. Fortunately not something we need to address right now!

Thanks for the review.

tahonermann committed:
UnsignedOrNone required following rebase for the discriminator override argument of ItaniumMangleContext::create().
